### PR TITLE
Wib 838/decrypt data upon unlock

### PR DESCRIPTION
**Pull request by nicosommi**
### What does this Pull Request do?
Adds the logic to decrypt the data from the buyer side.

### Are there points in the code the reviewer needs to double check?
Specially the decrypt job.

### Why was this Pull Request needed?
Feature WIB-838

### Does this Pull Request meet the acceptance criteria?
- [X] I have read the [Contribution guidelines](https://github.com/wibsonorg/buyer-sdk/CONTRIBUTING.md).
- [X] I have read the [Code of Conduct](https://github.com/wibsonorg/buyer-sdk/CODE_OF_CONDUCT.md).
- [X] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.

### CODE REVIEW

File: buyer-api/src/blockchain/contractEventSubscribers/batchPaymentsSubscribers.js [hunk 1]
```diff
@@ -28,13 +28,23 @@ export const updateBuyerStats = async (
   );
 };
 
+/**
+ * @function storeLockingKeyHashByPayIndex Is triggered when
+ * the payment is registered,
+ * storing locally a relationship between the pay index and
+ * the payment transaction hash
+ * @typedef PaymentRegisteredEventValues
```
Comment (Agus) [inline buyer-api/src/blockchain/contractEventSubscribers/batchPaymentsSubscribers.js]: I think that `typedef` should go outside the jsdocs of the functions, and then use it there, right?

File: buyer-api/src/blockchain/contractEventSubscribers/batchPaymentsSubscribers.js [hunk 2]
```diff
@@ -28,13 +28,23 @@ export const updateBuyerStats = async (
...
+ * @property {number} payIndex Pay index
+ * @param {UnlockEventValues} eventValues The values emmited by the BatPay PaymentRegistered event
+ */
 export const storeLockingKeyHashByPayIndex = async ({ payIndex }, { transactionHash }) =>
```
Comment (Agus) [inline buyer-api/src/blockchain/contractEventSubscribers/batchPaymentsSubscribers.js]: Update docs with two parameters, and there's no declaration of type `UnlockEventValues`

File: buyer-api/src/blockchain/contractEventSubscribers/batchPaymentsSubscribers.js [hunk 3]
```diff
@@ -28,13 +28,23 @@ export const updateBuyerStats = async (
...
   paymentsTransactionHashes.store(payIndex, transactionHash);
 
 /**
  * @function decryptSellerKeys Is triggered when the payment to the seller is unlocked
  * @typedef PaymentUnlockedEventValues
- * @property {string} buyer The buyer that created the DataOrder
+ * @property {number} payIndex Pay index
+ * @property {string} key Master key to decrypt seller keys
  * @param {UnlockEventValues} eventValues The values emmited by the BatPay PaymentUnlocked event
```
Comment (Agus) [inline buyer-api/src/blockchain/contractEventSubscribers/batchPaymentsSubscribers.js]: You could typedef `UnlockEventValues` outside, or just use those two properties as `parameter` here.

File: buyer-api/src/queues/decryptSellerKeys.js [hunk 4]
```diff
@@ -1,23 +1,52 @@
...
+  logger.debug('Took current order data object');
+  // TODO: double check, are we going to index totals by seller address?
+  // append reduced sellers data to total
+  const allBatchSellers = sellersWithData.reduce(
+    (acc, curr) => ({ ...acc, [curr.address]: curr.data }),
+    {},
+  );
+  logger.debug('All batch sellers', { allBatchSellers });
+  // put order object
+  putRawOrderData(orderId, { ...rawOrderData, ...allBatchSellers });
```
Comment (Agus) [inline buyer-api/src/queues/decryptSellerKeys.js]: This should be awaited I think

File: buyer-api/src/blockchain/contractEventSubscribers/batchPaymentsSubscribers.js (hunk 1 above)
Comment (nicosommi) [reply buyer-api/src/blockchain/contractEventSubscribers/batchPaymentsSubscribers.js]: vscode at least seem to work if I define the typedef just above using it within the function, check out this SS
Screen Shot 2019-07-04 at 18 50 03

**Comment by nicosommi** [comment]
The test coverage decay is mostly due to existing crypto functions on wibson-lib

File: buyer-api/src/blockchain/contractEventSubscribers/batchPaymentsSubscribers.js (hunk 1 above)
Comment (Agus) [reply buyer-api/src/blockchain/contractEventSubscribers/batchPaymentsSubscribers.js]: 😲 Interesting, ok!

File: buyer-api/src/blockchain/contractEventSubscribers/batchPaymentsSubscribers.js [hunk 5]
```diff
@@ -28,4 +33,44 @@ export const updateBuyerStats = async (
...
+ * @property {string} key Master key to decrypt seller keys
+
+ * @param {UnlockEventValues} event The values emmited by the BatPay PaymentUnlocked event
+ */
+export const decryptSellerKeys = async ({ payIndex, key: masterKey }) => {
+  const transactionHash = await paymentsTransactionHashes.safeFetch(payIndex, null);
+  if (!transactionHash) return; // Not our payment
+  const { lockingKeyHash } = await fetchTxData(transactionHash);
+  const notarizationId = await notarizationsPerLockingKeyHash.fetch(lockingKeyHash);
+  notarizations.update(notarizationId, {
```
Comment (nicoayala) [inline buyer-api/src/blockchain/contractEventSubscribers/batchPaymentsSubscribers.js]:
```suggestion
  await notarizations.update(notarizationId, {
```

File: buyer-api/src/blockchain/contractEventSubscribers/batchPaymentsSubscribers.js [hunk 6]
```diff
@@ -28,4 +33,44 @@ export const updateBuyerStats = async (
...
+    masterKey,
+  });
+  addDecryptJob({ notarizationId });
```
Comment (nicoayala) [inline buyer-api/src/blockchain/contractEventSubscribers/batchPaymentsSubscribers.js]:
```suggestion
  await addDecryptJob({ notarizationId });
```

File: buyer-api/src/queues/decryptSellerKeys.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  );
+  logger.debug('All batch sellers', { allBatchSellers });
+  // put order object
+  await putRawOrderData(orderId, { ...rawOrderData, ...allBatchSellers });
+  logger.debug('Bucket updated with decrypted data.');
+};
+
+queue.process('decrypt', decryptSellersKeysJobListener);
+queue.on('failed', ({ id, name, failedReason }) => {
+  logger.error(`N[${id}] :: ${name} :: Error thrown: ${failedReason} (will be retried)`);
```
Comment (nicoayala) [inline buyer-api/src/queues/decryptSellerKeys.js]:
```suggestion
  logger.error(`DSK[${id}] :: ${name} :: Error thrown: ${failedReason} (will be retried)`);
```

File: buyer-api/src/utils/wibson-lib/cryptography/ec-encription.js [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+  assert(
+    recoveredAddress.toLowerCase() === senderAddress.toLowerCase(),
+    'The message is not from the expected address',
+  );
+
+  return decryptedPayload.message;
+};
+
+export { encryptSignedMessage, decryptSignedMessage, encryptWithPublicKey, hashAndSignMessage };
```
Comment (nicoayala) [inline buyer-api/src/utils/wibson-lib/cryptography/ec-encription.js]: I think that `encryptSignedMessage` and `decryptSignedMessage` are also present in `buyer-api/src/utils/wibson-lib/cryptography/encription.js`. I would keep only those ones in the `encryption` module

**Comment by Agus** [comment]
@nicoayala Updated!

File: buyer-api/test/queues/decryptSellerKeys.test.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    notarizationId: job.data.notarizationId,
+  };
+  await addDecryptJob(params);
+  t.true(fakeQueue.add.calledOnceWithExactly('decrypt', params));
+});
+
+it('decryptSellersKeysJobListener works as expected', async (t) => {
+  await decryptSellersKeysJobListener(job);
+  t.true(notarizations.fetch.calledOnceWithExactly(job.data.notarizationId));
+  // t.true(notarizations.fetch.calledOnceWithExactly(job.notarizationId));
```
Comment (nicoayala) [inline buyer-api/test/queues/decryptSellerKeys.test.js]:
```suggestion

```

**Pull request merged**